snap/quota/quota.go

```diff
@@ -257,18 +259,39 @@ func ResolveCrossReferences(grps map[string]*Group) error {
 
 // tree recursively returns all of the sub-groups of the group and the group
 // itself.
-func (grp *Group) tree() []*Group {
-	// TODO: should we be paranoid about circular references and
-	// recursion here?
```

Comment:
I think recursion is still a problem? Should we leave the TODO for now? 

I think we should limit the depth to a sane number to avoid DOS but it's fine to leave it for later (and/or we could also later refactor this helper not to use recursion).

Reply:
Sorry can you elaborate on the problem? Do you just not like the recursion at all? I think that we properly detect cycles in the tree here, if you have an example of a cycle that is not detected then I'd be keen to see it, but I think that the code is good here for now. Also consider the fact that this is paranoid code, this will never be used on user input directly, just on what we have in state

Reply:
I wonder if systemd itself has a limit on slice nesting? anyway the filesystem will also be unhappy at some point because of the slice filename length. We might want to impose a nesting limit ourselves before we hit those others

Reply:
to be clear I put that TODO there, and I was more worried about the combination of the two than recursion in itself but see my other comment

Reply:
so should I impose that limit now in this PR or leave a TODO?

Reply:
Ah sorry, I didn't realize it's only used for what's already in the state, makes sense so that's fine here. And yes, the cycle detection looks good.

I'm just wondering if we do control and limit the nesting depth when creating subgroups (maybe I missed it in already landed PRs) and if we should do it to avoid hitting stack limit if e.g. someone is scripting this and makes a mistake that creates crazy number of nested groups.

Reply:
ok I left some TODOs